nimare/tests/test_estimator_performance.py

```diff
@@ -0,0 +1,268 @@
+
+    ################################################################
+    # CHECK IF META/KERNEL/CORRECTOR COMBINATION PERFORMS ADEQUATELY
+    ################################################################
+    # The below combinations cannot even detect
+    # significance at the ground truth foci voxel
+    # locations
+    if (
+        (
+            isinstance(meta, mkda.MKDADensity)
+            and isinstance(kern, kernel.KDAKernel)
+            and corr.method != "montecarlo"
+        )
+        or (
+            isinstance(meta, mkda.MKDADensity)
+            and isinstance(kern, kernel.MKDAKernel)
+            and corr.method != "montecarlo"
+        )
+        or (
+            isinstance(meta, ale.ALE)
+            and isinstance(kern, kernel.MKDAKernel)
+            and corr.method == "montecarlo"
+        )
+    ):
+        good_performance = False
+    else:
+        good_performance = True
+    mask = cres.masker.mask_img
+    ground_truth_foci_ijks = [tuple(mm2vox(focus, mask.affine)) for focus in ground_truth_foci]
+    # reformat coordinate indices to index p_values_data
+    gtf_idx = [
+        [ground_truth_foci_ijks[i][j] for i in range(len(ground_truth_foci_ijks))]
+        for j in range(3)
+    ]
+
+    p_values_volume = cres.masker.inverse_transform(p_values_data).get_fdata()
+    # the ground truth foci should be significant at minimum
+    best_chance_p_values = p_values_volume[gtf_idx]
+    assert all(best_chance_p_values < ALPHA) == good_performance
+
+    # double the radius of what is expected to be significant
+    r = fwhm
+    # create an expectation of significant/non-significant regions
+    sig_map, nonsig_map = _create_signal_mask(mask, ground_truth_foci_ijks, r=r)
+    sig_idx = cres.masker.transform(sig_map).astype(bool).squeeze()
+    nonsig_idx = cres.masker.transform(nonsig_map).astype(bool).squeeze()
+    # while this combination passes the basic
+    # test of significance at the ground truth foci voxels,
+    # it does not meet the criteria for 50% power
+    # around the ground truth foci
+    if (
+        isinstance(meta, mkda.MKDADensity)
+        and isinstance(kern, kernel.MKDAKernel)
+        and corr.method == "montecarlo"
+    ) or (
+        isinstance(meta, mkda.MKDADensity)
+        and isinstance(kern, kernel.KDAKernel)
+        and corr.method == "montecarlo"
+    ):
+        good_performance = False
+
+    # assert that at least 50% of voxels surrounding the foci
+    # are significant at alpha = .05
+    observed_sig = p_values_data[sig_idx] < ALPHA
+    observed_sig_perc = observed_sig.sum() / len(observed_sig)
+    assert (observed_sig_perc >= 0.5) == good_performance
+
+    # assert that more than 95% of voxels farther away
+    # from foci are nonsignificant at alpha = 0.05
+    observed_nonsig = p_values_data[nonsig_idx] > ALPHA
+    observed_nonsig_perc = observed_nonsig.sum() / len(observed_nonsig)
+    assert observed_nonsig_perc >= BETA
```

Comment:
Have not read very carefully, but unless `ALPHA` is set way lower than 0.05, I don't think this is a safe assertion... in the long run I think `observed_nonsig_perc` should be equal to what you're calling `BETA` (the true negative rate), but in the short run it could be lower than that. So probably we want to check that it's reasonably close to the true negative rate, where "reasonably close" can probably be defined via trial and error.

Reply:
Oh, unless this is still testing with 0 noise coordinates. If that's the case, then I think this is probably fine—because unless I'm mistaken, we should get tiny p-values everywhere (because there's no noise, so the p-values should be either 0 or the smallest possible permutation-based value). But we should definitely include at least one test that injects noise into the simulated data, so we can make sure that the false positive rate is maintained at a sane level outside areas where there's signal.

Reply:
I've injected noise foci into the test, and it looks like reasonable output with the ALE/ALEKernel, but the ALE/MKDAKernel performs poorly (one interesting note about the poor performance is that ALE/MKDAKernel appears to perform worse with the addition of more noise foci).